### PR TITLE
feat(executor): Auto create s3 bucket if not present. Closes #3586 

### LINKED ISSUE #3586: Create default S3 bucket if not present
# Summary

When writing output S3 artifacts to a bucket (e.g. default bucket), if the specified bucket doesn't exist, the execution fails. A message is shown in the Argo UI saying: `failed to save outputs: timed out waiting for the condition`.

I propose we add the option to enable the auto-creation of the required S3 bucket if it does not exist.

# Motivation

In a newly bootstrapped environment, the S3 bucket might not exist initially.

# Proposal

In the Save method in s3.go, check if the bucket exists and create it if it doesn't. (Could make it configurable via Config Map).

https://github.com/argoproj/argo/blob/cdc935ae76b3d7cc50a486695b40ff2f647b49bc/workflow/artifacts/s3/s3.go#L84

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by alexcapras**
Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

Closes #3586

### CODE REVIEW

**Comment by alexec** [comment]
Assigned to @jessesuen to approve proto changes.

**Comment by alexec** [comment]
@alexcapras can you please link to the issue and previous PR?

**Comment by alexcapras** [comment]
> @alexcapras can you please link to the issue and previous PR?

Issue: https://github.com/argoproj/argo/issues/3586
Previous PR: https://github.com/argoproj/argo/pull/4098/files

**Review by jessesuen — APPROVED** [review]
re-approving the proto changes.

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 1]
```diff
@@ -1650,6 +1650,21 @@ type S3Bucket struct {
...
+	CreateBucketIfNotPresent *CreateS3BucketOptions `json:"createBucketIfNotPresent,omitempty" protobuf:"bytes,9,opt,name=createBucketIfNotPresent"`
+}
+
+// CreateS3BucketOptions options used to determine automatic automatic bucket-creation process
+type CreateS3BucketOptions struct {
+	// Enabled tells the driver whether to attempt to create the S3 bucket or not
+	Enabled bool `json:"enabled,omitempty" protobuf:"varint,1,opt,name=enabled"`
+
+	// Region Bucket location
+	Region string `json:"region,omitempty" protobuf:"varint,2,opt,name=region"`
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:1664]: why do we need region - isn't still already specified by S3Bucket?

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 2]
```diff
@@ -1650,6 +1650,21 @@ type S3Bucket struct {
...
 	UseSDKCreds bool `json:"useSDKCreds,omitempty" protobuf:"varint,8,opt,name=useSDKCreds"`
+
+	// CreateBucketIfNotPresent tells the driver to attempt to create the S3 bucket for output artifacts, if it doesn't exist
+	CreateBucketIfNotPresent *CreateS3BucketOptions `json:"createBucketIfNotPresent,omitempty" protobuf:"bytes,9,opt,name=createBucketIfNotPresent"`
+}
+
+// CreateS3BucketOptions options used to determine automatic automatic bucket-creation process
+type CreateS3BucketOptions struct {
+	// Enabled tells the driver whether to attempt to create the S3 bucket or not
+	Enabled bool `json:"enabled,omitempty" protobuf:"varint,1,opt,name=enabled"`
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:1661]: Do we need `Enabled`? could the existence of `CreateS2BucketOptions` imply enabled?

File: test/e2e/testdata/output-artifact-with-s3-bucket-creation-disabled.yaml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      container:
+        image: argoproj/argosay:v1
+        command: [sh, -c]
+        args: ["sleep 1; cowsay hello world | tee /tmp/hello_world.txt"]
+      outputs:
+        artifacts:
+          - name: hello-art
+            path: /tmp/hello_world.txt
+            s3:
+              bucket: bucket-that-does-not-exist-1
```
Comment (alexec) [inline test/e2e/testdata/output-artifact-with-s3-bucket-creation-disabled.yaml:24]: looks good to me

File: test/e2e/testdata/output-artifact-with-s3-bucket-creation-disabled.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
+# This example demonstrates bucket-creation disabled for s3 output artifacts, when the target bucket does not exist.
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  generateName: output-artifact-with-s3-bucket-creation-disabled-
+spec:
+  entrypoint: output-artifact-with-s3-bucket-creation-disabled
```
Comment (alexec) [inline test/e2e/testdata/output-artifact-with-s3-bucket-creation-disabled.yaml:7]: minor - just change entrypoint to `main`

File: test/e2e/testdata/output-artifact-with-s3-bucket-creation-disabled.yaml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  templates:
+    - name: output-artifact-with-s3-bucket-creation-disabled
+      steps:
+        - - name: generate-artifact
+            template: whalesay
+
+    - name: whalesay
+      container:
+        image: argoproj/argosay:v1
```
Comment (alexec) [inline test/e2e/testdata/output-artifact-with-s3-bucket-creation-disabled.yaml:16]: can you please use `argosay:v2` - is it much less resource intensive

File: test/e2e/testdata/output-artifact-with-s3-bucket-creation-enabled.yaml [hunk 6]
```diff
@@ -0,0 +1,37 @@
+# This example demonstrates bucket-creation enabled for s3 output artifacts, when the target bucket does not exist.
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  generateName: output-artifact-with-s3-bucket-creation-enabled-
+  labels:
+    argo-e2e: true
+spec:
+  entrypoint: output-artifact-with-s3-bucket-creation-enabled
```
Comment (alexec) [inline test/e2e/testdata/output-artifact-with-s3-bucket-creation-enabled.yaml:9]: as above - rename to `main`

File: workflow/artifacts/s3/s3.go [hunk 7]
```diff
@@ -103,6 +105,28 @@ func (s3Driver *S3ArtifactDriver) Save(path string, outputArtifact *wfv1.Artifac
 				log.Warnf("Failed to test if %s is a directory: %v", path, err)
 				return false, nil
 			}
+
+			createBucketIfNotPresent := outputArtifact.S3.CreateBucketIfNotPresent
+			if createBucketIfNotPresent != nil && createBucketIfNotPresent.Enabled {
+				bucketExists, err := s3cli.BucketExists(outputArtifact.S3.Bucket)
```
Comment (alexec) [inline workflow/artifacts/s3/s3.go:111]: I suggest we just do a single `MakeBucket` and ignore duplicate errors. 50% of the network requests to go missing

**Review by alexec — CHANGES_REQUESTED** [review]
Changes requested.

**Comment by alexec** [comment]
Good work! You have some failing jobs that need remedy and you also need to sync with master and we are ready to merge!

**Review by alexec — APPROVED** [review]

File: test/e2e/testdata/output-artifact-with-s3-bucket-creation-disabled.yaml [hunk 8]
```diff
@@ -0,0 +1,34 @@
+# This example demonstrates bucket-creation disabled for s3 output artifacts, when the target bucket does not exist.
```
Comment (alexec) [inline test/e2e/testdata/output-artifact-with-s3-bucket-creation-disabled.yaml:1]: delete this file?

**Comment by alexec** [comment]
TestOutputArtifactS3BucketCreationEnabled failed?

**Comment by alexcapras** [comment]
> TestOutputArtifactS3BucketCreationEnabled failed?

@alexec Addressed this in my latest commit. Only drawback here is that if the user has enabled `createBucketIfNotPresent` and the `MakeBucket` call fails for a reason (other than the bucket already exists, e.g. insufficient perms), we will try to create it again in the ExponentialBackoff loop. (Because the bucketCreation will fail but we will still try to Put the artifact).

If this is not an issue then i think it should be ok now

**Comment by alexec** [comment]
Merged. This is currently targeted for v3.0. That'll be some time away, maybe Jan. I'll look into backporting it to v2.12

**Pull request merged**